### PR TITLE
Add Simplex noise UI operation

### PR DESCRIPTION
**Pull request by janper**
image

### CODE REVIEW

File: src/interpreter_funcs/mod.rs [hunk 1]
```diff
@@ -79,6 +81,7 @@ pub const FUNC_ID_VOXELIZE: FuncIdent = FuncIdent(8000);
 pub const FUNC_ID_BOOLEAN_INTERSECTION: FuncIdent = FuncIdent(8001);
 pub const FUNC_ID_BOOLEAN_UNION: FuncIdent = FuncIdent(8002);
 pub const FUNC_ID_BOOLEAN_DIFFERENCE: FuncIdent = FuncIdent(8003);
+pub const FUNC_ID_VOXEL_NOISE: FuncIdent = FuncIdent(8004);
```
Comment (yanchith) [inline src/interpreter_funcs/mod.rs:86]: Will need to resolve conflicts with master here

File: src/mesh/voxel_cloud.rs [hunk 2]
```diff
@@ -1020,6 +1020,33 @@ impl ScalarField {
...
+
+        let simplex = OpenSimplex::new();
+
+        for (one_dimensional, voxel) in self.voxels.iter_mut().enumerate() {
+            let absolute_coordinate = one_dimensional_to_absolute_voxel_coordinate(
+                one_dimensional,
+                &self.block_start,
+                &self.block_dimensions,
+            );
+            let noise_scale_f64 = noise_scale as f64;
```
Comment (yanchith) [inline src/mesh/voxel_cloud.rs]: `f64::from`

File: src/mesh/voxel_cloud.rs [hunk 3]
```diff
@@ -1020,6 +1020,33 @@ impl ScalarField {
...
+
+            let noise_value = simplex.get([
+                absolute_coordinate.x as f64 * noise_scale_f64,
+                absolute_coordinate.y as f64 * noise_scale_f64,
+                absolute_coordinate.z as f64 * noise_scale_f64,
+                time_offset as f64,
```
Comment (yanchith) [inline src/mesh/voxel_cloud.rs]: `f64::from` everywhere

File: src/interpreter_funcs/voxel_noise.rs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                    max_value: None,
+                    default_value_x: Some(1.0),
+                    default_value_y: Some(1.0),
+                    default_value_z: Some(1.0),
+                }),
+                optional: false,
+            },
+            ParamInfo {
+                name: "Voxel Size",
+                description: "Size of a single cell in the regular three-dimensional voxel grid.\n\
```
Comment (yanchith) [inline src/interpreter_funcs/voxel_noise.rs:89]: Two newlines here?

File: src/interpreter_funcs/voxel_noise.rs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                High values produce coarser results, low values may increase precision but produce \
+                heavier geometry that significantly affect performance. Too high values produce \
```
Comment (yanchith) [inline src/interpreter_funcs/voxel_noise.rs]: affects

File: src/interpreter_funcs/voxel_noise.rs [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+                    max_value: None,
+                }),
+                optional: false,
+            },
+            ParamInfo {
+                name: "Time offset",
+                description: "",
+                refinement: ParamRefinement::Float(FloatParamRefinement {
+                    default_value: Some(1.0),
+                    min_value: Some(f32::MIN_POSITIVE),
```
Comment (yanchith) [inline src/interpreter_funcs/voxel_noise.rs]: Why not zero? For 4D noise it is a dimension like any other.

File: src/interpreter_funcs/voxel_noise.rs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        let volume_range_raw = args[5].unwrap_float2();
+        let error_if_large = args[6].unwrap_boolean();
+        let analyze_bbox = args[7].unwrap_boolean();
+        let analyze_mesh = args[8].unwrap_boolean();
+
+        let meshing_range = volume_range_raw[0]..=volume_range_raw[1];
+
+        if voxel_dimensions
+            .iter()
+            .any(|dimension| approx::relative_eq!(*dimension, 0.0))
```
Comment (yanchith) [inline src/interpreter_funcs/voxel_noise.rs]: Account for negative values

File: src/interpreter_funcs/voxel_noise.rs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+            // The equation doesn't take rounding into consideration, hence the
+            // arbitrary multiplication by 1.1.
+            let error = FuncError::new(FuncVoxelNoiseError::TooManyVoxels(
+                VOXEL_COUNT_THRESHOLD,
+                vx * 1.1,
+                vy * 1.1,
+                vz * 1.1,
+            ));
+            log(LogMessage::error(format!("Error: {}", error)));
+            return Err(error);
```
Comment (yanchith) [inline src/interpreter_funcs/voxel_noise.rs]: I believe this has been extracted somewhere already?

File: src/interpreter_funcs/voxel_noise.rs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+        scalar_field.fill_with_noise(noise_scale, time_offset);
+
+        if !scalar_field.contains_voxels_within_range(&meshing_range) {
+            let error = FuncError::new(FuncVoxelNoiseError::EmptyScalarField);
+            log(LogMessage::error(format!("Error: {}", error)));
+            return Err(error);
+        }
+
+        match scalar_field.to_mesh(&meshing_range) {
```
Comment (yanchith) [inline src/interpreter_funcs/voxel_noise.rs]: We are missing marching cubes here.

**Review by yanchith — CHANGES_REQUESTED** [review]
Noise! See inline comments. Mostly the PR is a bit out of date, but nothing major.

**Comment by janper** [comment]
ready for another review

**Review by yanchith — APPROVED** [review]

**Pull request merged**